libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,476 @@
+         std::ceil(maxBpm * barDuration / 60), minBeatsPerBar, maxBeatsPerBar);
+      std::vector<int> possibleBeatsPerBar(maxBpb - minBpb + 1);
+      std::iota(possibleBeatsPerBar.begin(), possibleBeatsPerBar.end(), minBpb);
+      std::for_each(
+         possibleBeatsPerBar.begin(), possibleBeatsPerBar.end(),
+         [&](int beatsPerBar) {
+            std::for_each(
+               possibleTatumsPerBeat.begin(), possibleTatumsPerBeat.end(),
+               [&](double tatumsPerBeat) {
+                  const auto tatumsPerBar = beatsPerBar * tatumsPerBeat;
+                  const int numTatums = tatumsPerBar * numBars;
+                  const auto tatumRate = numTatums / audioFileDuration * 60;
+                  if (
+                     IsRound(tatumsPerBar) && minTatumsPerMinute < tatumRate &&
+                     tatumRate < maxTatumsPerMinute)
+                     possibleDivHierarchies[numTatums].push_back(
+                        BarDivision { numBars, beatsPerBar });
+               });
+         });
+   });
+   return possibleDivHierarchies;
+}
+
+int GetOnsetLag(const std::vector<float>& odf, int numTatums)
+{
+   // Understandably, the first onset of a loop recording isn't typically
+   // centered on time 0, or the would be a click at the onset. The entire
+   // recording is therefore likely to have a small lag. It is important that we
+   // take this lag into account for the quantization distance measure.
+   // The code below is equivalent to cross-correlating the odf with a pulse
+   // train of frequency `numTatums / odf.size()`. We take the position of the
+   // first peak to be the lag.
+   const auto odfSamplesPerDiv = 1. * odf.size() / numTatums;
+   auto max = std::numeric_limits<float>::lowest();
+   auto lag = 0;
+   std::vector<int> tatumIndices(numTatums);
+   std::iota(tatumIndices.begin(), tatumIndices.end(), 0);
+   while (true)
+   {
+      const auto val = std::accumulate(
+         tatumIndices.begin(), tatumIndices.end(), 0.f, [&](float sum, int i) {
+            const int j = std::round(i * odfSamplesPerDiv) + lag;
+            return sum + (j < odf.size() ? odf[j] : 0.f);
+         });
+      if (val < max)
+         break;
+      max = val;
+      ++lag;
+   }
+   return lag - 1;
+}
+
+// This is the fundament of the algorithm. It gives a weighted average of the
+// normalized distance between ODF peaks and the closest tatum. The weights are
+// the ODF values at the peaks. The distance is normalized by the tatum
+// duration, so that the final value ranges between 0 and 1.
+// This can be seen as a performance accuracy measure. The higher the tempo, the
+// more accurate the performer has to be for this distance to stay low. In
+// general, this is desired. We could nevertheless consider introducing a slight
+// tolerance which increases with the tatum rate, as very fast rhythmic patterns
+// recorded without post-editing still tend to achieve lower scores.
```

Comment:
This comment makes me think of application of a varying stretch to correct irregularities of rhythm... have you thought of that?

Reply:
This could be an application of the result of this analysis indeed. We could offer the user the possibility to automatically warp the onsets to the nearest tatum. Generalizing our integration of the time-stretching from StaffPad may not be that tricky, but there would be real difficulties related to waveform painting. Or maybe the painting would be so subtle we wouldn't even need to address it ... then pencil and brush would also have to be disabled for time-warped clips, but UX-wise that wouldn't be too bad either.